File: frontend/src/lib/rag.ts
import { useMutation, useQuery } from "@tanstack/react-query";
import { api } from "./api";

export interface RagSource {
  title: string;
  source: string;
  url: string;
  score: number;
}

export interface RagAnswer {
  answer: string;
  sources: RagSource[];
  grounded: boolean;
  llm: string;
}

export interface RagStatus {
  indexed_chunks: number;
  topics: string[];
  embedder: string | null;
}

export function useRagStatus() {
  return useQuery<RagStatus>({
    queryKey: ["rag-status"],
    queryFn: async () => (await api.get<RagStatus>("/rag/status")).data,
  });
}

export function useRagAsk() {
  return useMutation({
    mutationFn: async (payload: { question: string; topic?: string }) =>
      (await api.post<RagAnswer>("/rag/ask", payload)).data,
  });
}

File: backend/app/schemas/rag.py
from pydantic import BaseModel, Field


class RagQuery(BaseModel):
    question: str = Field(min_length=1, max_length=1000)
    topic: str | None = None
    k: int = Field(default=4, ge=1, le=8)


class RagSource(BaseModel):
    title: str
    source: str
    url: str
    score: float


class RagAnswer(BaseModel):
    answer: str
    sources: list[RagSource]
    grounded: bool
    llm: str


class RagStatus(BaseModel):
    indexed_chunks: int
    topics: list[str]
    embedder: str | None
